homeassistant/helpers/restore_state.py

```diff
@@ -1,98 +1,113 @@
 
-import async_timeout
+DATA_RESTORE_STORAGE = 'restore_state_store'
+DATA_RESTORE_CACHE_TASK = 'restore_state_store_task'
 
-from homeassistant.core import HomeAssistant, CoreState, callback
-from homeassistant.const import EVENT_HOMEASSISTANT_START
-from homeassistant.loader import bind_hass
-from homeassistant.components.history import get_states, last_recorder_run
-from homeassistant.components.recorder import (
-    wait_connection_ready, DOMAIN as _RECORDER)
-import homeassistant.util.dt as dt_util
-
-RECORDER_TIMEOUT = 10
-DATA_RESTORE_CACHE = 'restore_state_cache'
-_LOCK = 'restore_lock'
 _LOGGER = logging.getLogger(__name__)
 
+STORAGE_KEY = 'core.restore_state'
+STORAGE_VERSION = 1
+
+# How long between periodically saving the current states to disk
+STATE_DUMP_INTERVAL = timedelta(minutes=15)
 
-def _load_restore_cache(hass: HomeAssistant):
-    """Load the restore cache to be used by other components."""
+
+@callback
+def async_setup(hass: HomeAssistant) -> None:
+    """Set up the event listeners for state restoration."""
     @callback
-    def remove_cache(event):
-        """Remove the states cache."""
-        hass.data.pop(DATA_RESTORE_CACHE, None)
+    def async_add_dump_states_job(*args):
+        """Set up the restore state listeners."""
+        hass.async_create_task(async_dump_states(hass))
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, remove_cache)
+    @callback
+    def async_setup_restore_state(*args):
+        """Set up the restore state listeners."""
+        # Dump the initial states now. This helps minimize the risk of having
+        # old states loaded by overwritting the last states once home assistant
+        # has started.
+        async_add_dump_states_job()
 
-    last_run = last_recorder_run(hass)
+        # Dump states periodically
+        async_track_time_interval(
+            hass, async_add_dump_states_job, STATE_DUMP_INTERVAL)
 
-    if last_run is None or last_run.end is None:
-        _LOGGER.debug('Not creating cache - no suitable last run found: %s',
-                      last_run)
-        hass.data[DATA_RESTORE_CACHE] = {}
-        return
+        # Dump states when stopping hass
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP, async_add_dump_states_job)
 
-    last_end_time = last_run.end - timedelta(seconds=1)
-    # Unfortunately the recorder_run model do not return offset-aware time
-    last_end_time = last_end_time.replace(tzinfo=dt_util.UTC)
-    _LOGGER.debug("Last run: %s - %s", last_run.start, last_end_time)
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_START, async_setup_restore_state)
 
-    states = get_states(hass, last_end_time, run=last_run)
 
-    # Cache the states
-    hass.data[DATA_RESTORE_CACHE] = {
-        state.entity_id: state for state in states}
-    _LOGGER.debug('Created cache with %s', list(hass.data[DATA_RESTORE_CACHE]))
+async def async_get_restore_cache(hass: HomeAssistant) -> Dict[str, State]:
```

Comment:
We should rewrite this as a class that holds on to the store. It will make the code easier to reason about and will make future enhancements like keeping track of states that requested from the cache.